src/xdsl/irdl.py

```diff
@@ -649,12 +736,61 @@ def irdl_build_attribute(irdl_def: AttrConstraint, result: Any) -> Attribute:
     raise Exception(f"builder expected an attribute, got {result}")
 
 
+def irdl_build_arg_list(arg_type: ArgType,
+                        args: Sequence[Any],
+                        arg_defs: Sequence[tuple[str, Any]],
+                        error_prefix: str = "") -> tuple[list[Any], list[int]]:
+    """Build a list of arguments (operands, results, regions)"""
+
+    def build_arg(arg_def: Any, arg: Any) -> Any:
+        """Build a single argument."""
+        if arg_type == ArgType.OPERAND:
+            return SSAValue.get(arg)
+        elif arg_type == ArgType.RESULT:
+            assert isinstance(arg_def, ResultDef)
+            return irdl_build_attribute(arg_def.constr, arg)
+        elif arg_type == ArgType.REGION:
+            assert isinstance(arg_def, RegionDef)
+            return Region.get(arg)
+        else:
+            assert False, "Unknown ArgType value"
+
+    if len(args) != len(arg_defs):
+        raise ValueError(f"expected {len(arg_defs)} {arg_type.get_name()}, "
+                         f"but got {len(args)}")
+
+    res = list[Any]()
+    arg_sizes = list[int]()
+
+    for arg_idx, ((arg_name, arg_def), arg) in enumerate(zip(arg_defs, args)):
+        if isinstance(arg_def, VariadicDef):
+            if not isinstance(arg, list):
+                raise ValueError(error_prefix +
+                                 f"variadic {arg_type} {arg_idx} '{arg_name}' "
+                                 f" expects a list, but got {arg}")
+            arg = cast(list[Any], arg)
+
+            # Check we have at most one argument for optional defintions.
+            if isinstance(arg_def, OptionalDef) and len(arg) > 1:
+                raise ValueError(error_prefix +
+                                 f"optional {arg_type} {arg_idx} '{arg_name}' "
+                                 "expects a list of size at most 1, but "
+                                 f"got a list of size {len(arg)}")
+
+            res.extend([build_arg(arg_def, arg_arg) for arg_arg in arg])
+            arg_sizes.append(len(arg))
+        else:
+            res.append(build_arg(arg_def, arg))
+            arg_sizes.append(1)
+    return res, arg_sizes
+
+
 def irdl_op_builder(cls: type[_OpT], op_def: OpDef,
-                    operands: list[SSAValue | Operation
-                                   | list[SSAValue | Operation]],
-                    res_types: list[Any | list[Any]],
+                    operands: Sequence[SSAValue | Operation
```

Comment:
Why use Sequence here while we are using list everywhere else? I see that a list is a Sequence of some sort, but this feels inconsistent with the rest of the code.

Reply:
So in this function, the reasoning is that we don't care what kind of Sequence we will have for operands, but we do care that its elements are sometimes lists rather than sequences. The reason is that we are using `isinstance(..., list)`, to check if we have a list or something else.
I made a quick pass again on the code, and I spotted two lists that should have been Sequence though, so that's fixed.